code/datums/mapgen/MarsGenerator.dm

```diff
@@ -0,0 +1,201 @@
+		BIOME_LOWMEDIUM_HUMIDITY = /datum/biome/mars/duststorm,
+		BIOME_HIGHMEDIUM_HUMIDITY = /datum/biome/mars/duststorm,
+		BIOME_HIGH_HUMIDITY = /datum/biome/mars/martian_area
+		)
+	)
+	///Used to select "zoom" level into the perlin noise, higher numbers result in slower transitions
+	var/perlin_zoom = 65
+
+ ///Seeds the rust-g perlin noise with a random number.
+/datum/map_generator/mars_generator/generate_terrain(var/list/turfs, var/reuse_seed)
+	. = ..()
+	var/height_seed = seeds[1]
+	var/humidity_seed = seeds[2]
+	var/heat_seed = seeds[3]
+
+	for(var/t in turfs) //Go through all the turfs and generate them
+		var/turf/gen_turf = t
+		var/drift_x = (gen_turf.x + rand(-MARS_BIOME_RANDOM_SQUARE_DRIFT, MARS_BIOME_RANDOM_SQUARE_DRIFT)) / perlin_zoom
+		var/drift_y = (gen_turf.y + rand(-MARS_BIOME_RANDOM_SQUARE_DRIFT, MARS_BIOME_RANDOM_SQUARE_DRIFT)) / perlin_zoom
+
+		var/height = text2num(rustg_noise_get_at_coordinates("[height_seed]", "[drift_x]", "[drift_y]"))
+
+		var/datum/biome/selected_biome
+		if(height <= 0.85) //If height is less than 0.85, we generate biomes based on the heat and humidity of the area.
+			var/humidity = text2num(rustg_noise_get_at_coordinates("[humidity_seed]", "[drift_x]", "[drift_y]"))
+			var/heat = text2num(rustg_noise_get_at_coordinates("[heat_seed]", "[drift_x]", "[drift_y]"))
+			var/heat_level //Type of heat zone we're in LOW-MEDIUM-HIGH
+			var/humidity_level  //Type of humidity zone we're in LOW-MEDIUM-HIGH
+
+			switch(heat)
+				if(0 to 0.25)
+					heat_level = BIOME_LOW_HEAT
+				if(0.25 to 0.5)
+					heat_level = BIOME_LOWMEDIUM_HEAT
+				if(0.5 to 0.75)
+					heat_level = BIOME_HIGHMEDIUM_HEAT
+				if(0.75 to 1)
+					heat_level = BIOME_HIGH_HEAT
+			switch(humidity)
+				if(0 to 0.25)
+					humidity_level = BIOME_LOW_HUMIDITY
+				if(0.25 to 0.5)
+					humidity_level = BIOME_LOWMEDIUM_HUMIDITY
+				if(0.5 to 0.75)
+					humidity_level = BIOME_HIGHMEDIUM_HUMIDITY
+				if(0.75 to 1)
+					humidity_level = BIOME_HIGH_HUMIDITY
+			selected_biome = possible_biomes[heat_level][humidity_level]
+		else //Over 0.85; It's the abyss
+			selected_biome = /datum/biome/mars/minable
+		selected_biome = biomes[selected_biome]
+		selected_biome.generate_turf(gen_turf)
+
+		if (current_state >= GAME_STATE_PLAYING)
+			LAGCHECK(LAG_LOW)
+		else
+			LAGCHECK(LAG_HIGH)
+
+/turf/unsimulated/floor/setpieces/martian/station_duststorm
+
+	New()
+		src.rocks = prob(10)
+		..()
+
+	Entered(atom/movable/O)
+		..()
+		if (ishuman(O))
+			if(!ON_COOLDOWN(O,"mars_duststorm", rand(5 SECONDS, 15 SECONDS)))
+				var/mob/living/jerk = O
+				if (!isdead(jerk))
+					if((istype(jerk:wear_suit, /obj/item/clothing/suit/armor/mars))&&(istype(jerk:head, /obj/item/clothing/head/helmet/mars))) return
+					step(jerk,EAST)
+					if(jerk.protected_from_space()) // Be kind around station...
+						return
+					random_brute_damage(jerk, 20, checkarmor=TRUE) // Allow armor to resist
+					jerk.changeStatus("weakened", 5 SECONDS)
```

Comment:
possible alternative. If you don't like this then instead just be sure to add a `force_laydown_standup` after the weakened effect is addded 
```suggestion
					jerk.do_disorient(stamina_damage = 100, weakened = 3 SECONDS, disorient = 5 SECOND)
```